Makefile

```diff
@@ -262,30 +268,44 @@ test: $(ROM)
 $(ROM): $(ELF)
 	$(ELF2ROM) -cic 6105 $< $@
 
-$(ELF): $(TEXTURE_FILES_OUT) $(ASSET_FILES_OUT) $(O_FILES) $(OVL_RELOC_FILES) build/ldscript.txt build/undefined_syms.txt
-	$(LD) -T build/undefined_syms.txt -T build/ldscript.txt --no-check-sections --accept-unknown-input-arch --emit-relocs -Map build/z64.map -o $@
+$(ELF): $(SEGMENT_OBJECTS) $(SEGMENT_RELOCS) build/ldscript.txt build/undefined_syms.txt
+	$(LD) -T build/undefined_syms.txt -T build/ldscript.txt --emit-relocs -Map build/z64.map -o $@
 
-## Order-only prerequisites 
-# These ensure e.g. the O_FILES are built before the OVL_RELOC_FILES.
-# The intermediate phony targets avoid quadratically-many dependencies between the targets and prerequisites.
+build/undefined_syms.txt: undefined_syms.txt
+	$(CPP) $(CPPFLAGS) $< > $@
 
-o_files: $(O_FILES)
-$(OVL_RELOC_FILES): | o_files
+build/$(SPEC): $(SPEC)
+	$(CPP) $(CPPFLAGS) $< > $@
 
-asset_files: $(TEXTURE_FILES_OUT) $(ASSET_FILES_OUT)
-$(O_FILES): | asset_files
+build/ldscript.txt: build/$(SPEC)
+	$(MKLDSCRIPT) -r $< $(SEGMENT_DIR) $@
 
-.PHONY: o_files asset_files
+$(SEGMENT_SCRIPTS): build/$(SPEC)
+$(SEGMENT_SCRIPTS): %.ld: %.d
```

Comment:
It's because mkldscript generates both the dependency file and the script.

Reply:
I don't understand, they are indeed both generated at the same time, so why would one depend on the other?
Also, typically (always) `.ld`/`.d` will be newer than `build/$(SPEC)` and the `.ld`s already depend on `build/$(SPEC)`, so the `.d` would also be up to date?

Also unrelated to my question but even if this is needed, it assumes the `.ld` is written before the `.d`, or "not too long" after (whatever make's or the system's definition of "not too long" is)
(currently `mkldscript` writes the `.ld` after the `.d` time-wise)

Reply:
If you remove this dependency relation, then make would no longer know how to generate `.ld`. You would have to provide a recipe for it, which would be the same as `.d`. This would cause two invocations of mkldscript to occur, meaning both files would be generated twice, a waste of time. Having `.d` be newer than `.ld` is not a problem, because `.ld` does not have a recipe and thus will not be regenerated, even if it is considered "out of date" in a sense.